src/components/CreateVmWizard/helpers.js

```diff
@@ -1,22 +1,67 @@
-import { createTemplateList } from '_/components/utils'
+import { createIsoList, createTemplateList } from '_/components/utils'
 import timezones from '_/components/utils/timezones.json'
 import { isWindows } from '_/helpers'
 
-const handleClusterIdChange = (clusterId, blankTemplateId, { templates, clusters, defaultValues, operatingSystems, defaultGeneralTimezone, defaultWindowsTimezone }) => {
+const handleClusterIdChange = (clusterId, { blankTemplateId, defaultValues, clusters, templates, operatingSystems, storageDomains, defaultGeneralTimezone, defaultWindowsTimezone }) => {
   let changes = {}
   const templateList = clusterId ? createTemplateList(templates, clusterId) : []
   changes.dataCenterId = clusterId ? clusters.getIn([clusterId, 'dataCenterId']) : undefined
   changes.clusterId = clusterId
-  changes.provisionSource =
-    (templateList.length === 1 && templateList[0].id === blankTemplateId)
-      ? 'iso'
-      : 'template'
-  changes.templateId = templateList.length === 1 ? blankTemplateId : undefined
-  changes = { ...changes, ...handleProvisionSourceChange(changes.provisionSource, defaultValues, { defaultGeneralTimezone, defaultWindowsTimezone, templates, operatingSystems }) }
+
+  if (clusterId === undefined) {
+    changes = { ...changes, ...handleProvisionSourceChange('template', { defaultValues, defaultGeneralTimezone, defaultWindowsTimezone, templates, operatingSystems }) }
+    return changes
+  }
+
+  if (templateList.length > 1) {
+    // cluster includes more than one non blank template
+    changes = {
+      ...changes,
+      ...handleProvisionSourceChange('template', { defaultValues, defaultGeneralTimezone, defaultWindowsTimezone, templates, operatingSystems }),
+    }
+    if (templateList.length === 2) {
+      // cluster includes only one non blank template
+      const selectedTemplate = templateList.find(t => t.id !== blankTemplateId)
+      changes = {
+        ...changes,
+        ...(selectedTemplate ? handleTemplateIdChange(selectedTemplate.id, defaultValues.optimizedFor, { templates, operatingSystems, defaultGeneralTimezone, defaultWindowsTimezone }) : {}),
+      }
+    }
+    return changes
+  }
+  if (templateList.length === 1 && templateList[0].id === blankTemplateId) {
+    // cluster includes only one template (Blank)
+    const isoList = createIsoList(storageDomains, changes.dataCenterId)
+
+    if (isoList.length > 0) {
+      // cluster includes at least one iso cd
+      changes = {
+        ...changes,
+        ...handleProvisionSourceChange('iso', { defaultValues, defaultGeneralTimezone, defaultWindowsTimezone, templates, operatingSystems }),
+      }
+      if (isoList.length === 1) {
+        // cluster includes only one iso cd
+        changes.isoImage = isoList[0].file.id
+      }
+    } else {
+      // cluster does not includes any non blank template or iso cd
```

Comment:
The cluster "includes" or "does not" include any non blank template here? Could you, please, change this comment to "cluster does not include any non blank template or iso cd"? Let's have the comments grammatically correct to prevent any potential misunderstandings ;) Thanks.

Reply:
You are right, I wasn't focused, I missed that little 's', Thanks :)

Reply:
> You are right, I wasn't focused, I missed that little 's', Thanks :)

looks ok now, so merging.